### PR TITLE
merkle: Move rehashing algorithm in

### PR DESCRIPTION
**Pull request by pavelkalinnikov**
This change moves the `rehasher` algorithm to `merkle` package, and refactors it
to be a single function.

Part of #2143

### CODE REVIEW

**Comment by pavelkalinnikov** [comment]
Might be convenient or not to review commit by commit.

File: merkle/log_proofs.go [hunk 1]
```diff
@@ -177,7 +176,7 @@ func proofNodes(index uint64, level uint, size uint64, rehash bool) []NodeFetch
 
 // Rehasher bundles the rehashing logic into a simple state machine
 type Rehasher struct {
-	Hasher     hashers.LogHasher
+	Hash       func(left, right []byte) []byte
```
Comment (Martin2112) [inline merkle/log_proofs.go]: Using the same name for functions and values. HashFn?

File: merkle/log_proofs.go [hunk 2]
```diff
@@ -178,7 +177,7 @@ func proofNodes(index uint64, level uint, size uint64, rehash bool) []NodeFetch
 type Rehasher struct {
 	Hash       func(left, right []byte) []byte
 	rehashing  bool
-	rehashNode tree.Node
+	rehashHash []byte
```
Comment (Martin2112) [inline merkle/log_proofs.go]: If other field is renamed this could just be 'hash'.

File: merkle/log_proofs.go [hunk 3]
```diff
@@ -173,55 +174,27 @@ func proofNodes(index uint64, level uint, size uint64, rehash bool) []NodeFetch
...
-	r.rehashHash = hash
-	r.rehashing = true
-}
-
-func (r *Rehasher) endRehashing() {
-	if r.rehashing {
-		r.proof = append(r.proof, r.rehashHash)
-		r.rehashing = false
+	cursor := 0
+	for i, ln := 0, len(h); i < ln; i, cursor = i+1, cursor+1 {
```
Comment (Martin2112) [inline merkle/log_proofs.go:191]: I think this readers of this block could benefit from a couple of brief comments or an example. e.g. at the top what is cursor and inside what's the terminating condition on the loop.

Comment (pavelkalinnikov) [reply merkle/log_proofs.go:194]: Done.

**Pull request merged**